src/library_openal.js

```diff
@@ -1,1622 +1,4591 @@
+            // Sample the current time as late as possible to mitigate drift
+            startTime = Math.max(startTime, src.context.audioCtx.currentTime);
+            audioSrc.start(startTime, startOffset);
+          } else if (typeof(audioSrc.noteOn) !== "undefined") {
+            startTime = Math.max(startTime, src.context.audioCtx.currentTime);
+            audioSrc.noteOn(startTime);
+#if OPENAL_DEBUG
+            if (offset > 0.0) {
+              Runtime.warnOnce("The current browser does not support AudioBufferSourceNode.start(when, offset); method, so cannot play back audio with an offset "+startOffset+" secs! Audio glitches will occur!");
             }
 #endif
           }
 #if OPENAL_DEBUG
           else {
-            Runtime.warnOnce('Unable to start AudioBufferSourceNode playback! Not supported by the browser?');
+            Runtime.warnOnce("Unable to start AudioBufferSourceNode playback! Not supported by the browser?");
           }
 
-          console.log('updateSource queuing buffer ' + i + ' for source ' + idx + ' at ' + startTime + ' (offset by ' + offset + ')');
+          console.log("scheduleSourceAudio() queuing buffer " + buf.id + " for source " + src.id + " at " + startTime + " (offset by " + startOffset + ")");
 #endif
+          audioSrc._startTime = startTime;
+          src.audioQueue.push(audioSrc);
+
+          startTime += duration;
         }
 
-        startTime = endTime;
+        startOffset = 0.0;
+        bufCursor++;
       }
     },
 
-    setSourceState: function setSourceState(src, state) {
-#if OPENAL_DEBUG
-      var idx = AL.srcIdBySrc(src);
-#endif
+    // Advance the state of a source forward to the current time
+    updateSourceTime: function(src) {
+      var currentTime = src.context.audioCtx.currentTime;
+      if (src.state !== 0x1012 /* AL_PLAYING */) {
+        return currentTime;
+      }
+
+      // if the start time is unset, determine it based on the current offset
+      if (!isFinite(src.bufStartTime)) {
+        src.bufStartTime = currentTime - src.bufOffset / src.playbackRate;
+        src.bufOffset = 0.0;
+      }
+
+      var nextStartTime = 0.0;
+      while (src.audioQueue.length) {
+        var audioSrc = src.audioQueue[0];
+        src.bufsProcessed += audioSrc._skipCount;
+        nextStartTime = audioSrc._startTime + audioSrc._duration; // n.b. audioSrc._duration already factors in playbackRate, so no divide by src.playbackRate on it.
+
+        if (currentTime < nextStartTime) {
+          break;
+        }
+
+        src.audioQueue.shift();
+        src.bufStartTime = nextStartTime;
+        src.bufOffset = 0.0;
+        src.bufsProcessed++;
+      }
+
+      var audioSrc = src.audioQueue[0];
+      if (audioSrc) {
+        src.bufOffset = (currentTime - audioSrc._startTime) * src.playbackRate;
+      } else {
+        var skipCount = 0;
+        // In the absence of scheduled buffers, fast-forward playback state to the current time
+        while (true) {
```

Comment:
The purpose of this loop is bring the openal impl's internal state back up to date to match the current time of the webaudio context. It's designed to handle arbitrarily long pauses that could be caused by any kind of blocking delay elsewhere in the code. If a looping buffer queue was started prior to a long pause, it could have looped any number of times in the interim, so we need to loop as many times as it takes to get to the correct playback states.

The primary termination condition is when the next computed buffer start time is in the future, which is calculated by repeatedly adding buffer durations, so the value should be monotonically increasing. A queue full of 0-length buffers can't cause an infinite loop since those are explicitly checked for and the loop aborts if the number of 0-length buffers (skipCount) becomes equal to the queue length.

Still, we could add some kind of emergency abort threshold just in case, if you feel it's necessary.